sdk/client/src/otel/index.ts

```diff
@@ -0,0 +1,319 @@
+		super(exporter)
+
+		this.tracingOrigins = config.tracingOrigins ?? false
+	}
+
+	onStart(): void {
+		// Do nothing. Types required us to implement this method.
+	}
+
+	onEnd(span: ReadableSpan): void {
+		if (typeof span.attributes['http.url'] === 'string') {
+			const shouldRecordNetworkRequest = shouldNetworkRequestBeTraced(
+				span.attributes['http.url'],
+				this.tracingOrigins,
+			)
+
+			if (!shouldRecordNetworkRequest) {
+				span.spanContext().traceFlags = 0 // prevents span from being recorded
+			}
+		}
+	}
+
+	onShutdown(): void {
+		// Do nothing. Types required us to implement this method.
+	}
+}
+
+export const getOtelProvider = (): WebTracerProvider => {
+	if (provider === undefined) {
+		throw new Error('OTEL provider not initialized')
+	}
+
+	return provider
+}
+
+const getSpanName = (
+	url: string,
+	method: string,
+	body: Request['body'] | BrowserXHR['_body'],
+) => {
+	let parsedBody
+	const pathname = new URL(url).pathname
+	let spanName = `${method} - ${pathname}`
+
+	try {
+		parsedBody = typeof body === 'string' ? JSON.parse(body) : body
+		if (parsedBody && parsedBody.operationName) {
+			spanName = `${parsedBody.operationName} (GraphQL: ${pathname})`
+		}
+	} catch {
+		// Ignore
+	}
+
+	return spanName
+}
+
+const enhanceSpanWithHttpRequestAttributes = (
+	span: api.Span,
+	body: Request['body'] | RequestInit['body'] | BrowserXHR['_body'],
+	headers:
+		| Headers
+		| string
+		| Request['headers']
+		| RequestInit['headers']
+		| ReturnType<XMLHttpRequest['getAllResponseHeaders']>,
+	networkRecordingOptions?: NetworkRecordingOptions,
+) => {
+	let parsedBody
+
+	try {
+		parsedBody = body ? JSON.parse(String(body)) : undefined
+	} catch {
+		// Ignore
+	}
+
+	if (parsedBody) {
+		try {
+			setObjectAttributes(span, parsedBody, 'http.request.body')
+
+			if (parsedBody.operationName) {
+				span.setAttributes({
+					'graphql.operation.name': parsedBody.operationName,
+				})
+			}
+		} catch {
+			// Ignore
+		}
+	}
+
+	const sanitizedHeaders = sanitizeHeaders(
+		networkRecordingOptions?.networkHeadersToRedact ?? [''],
+		headers as Headers,
+		networkRecordingOptions?.headerKeysToRecord,
+	)
+
+	span.setAttribute('highlight.type', 'http.request')
+	span.setAttribute('http.request.headers', JSON.stringify(sanitizedHeaders))
+}
+
+const enhanceSpanWithHttpResponseAttributes = (
+	span: api.Span,
+	body: Request['body'] | Response['body'] | XMLHttpRequest['responseText'],
+	headers: Headers | string,
+	networkRecordingOptions?: NetworkRecordingOptions,
+) => {
+	let parsedBody
+
+	try {
+		parsedBody = typeof body === 'string' ? JSON.parse(body) : body
+	} catch {
+		// Ignore
+	}
+
+	const recordedBody = getBodyThatShouldBeRecorded(
+		parsedBody,
+		networkRecordingOptions?.networkBodyKeysToRedact,
+		networkRecordingOptions?.bodyKeysToRecord,
+		headers as Headers,
+	)
+
+	// TODO: We don't seem to be able to access this sometimes. Figure out why.
```

Comment:
Need to look into this, but figured I'd handle in a follow up PR. It might require creating our own instrumentation and waiting until the requests are actually finished to submit spans so that we can get the full response data.

Reply:
interesting... could our fetch instrumentation ever collide with otel's?